### DISCUSSION TITLE
Proposal: Literal identifiers with backtick

### DISCUSSION THREAD

**Discussion by mdmcconnell** [antonmedv/expr#370] [General]
My application has a lot of identifiers that use natural language, things like "Section 1-2a", or "2015 Information Table".  Currently, expr can not parse these as  single token identifiers for various reasons.  Possible workarounds for me include:  make ad-hoc subsitutions to my identifiers, such as "Section_1_2a", though that will be messy and error-prone; create a literal('Section 1-2a') method, which would clutter expressions; use single quote 'Section 1-2a' and patch the tree to turn all string nodes into indentifiers, but I would lose the ability to use actual strings in my expressions (porbably ok in my case).

I would like to know if expr would consider allowing backtick to enclose literal identifers, so the above `Section 1-2a` would be parsed as a single identifier node.  As backtick is now an illegal character I can not see backwards compatibility problems - though enabling this feature could be wrapped in an option to ensure safety.  Python has deprecated backtick, so a disadvantage could be some divergence from Python-like syntax.  Or possibly backtick is reserved for some future use like binary literals.  Any thoughts?

**Comment by antonmedv** [comment]
Can you show an example of what it may look like? Is it for variable access from env?

**Reply by mdmcconnell** [reply]
Yes, correct my use is for variable access from env (I am using a simple map[string]struct for env - in go, any string could be a key).  There would also be the possibility of function name literals and others; maybe that is of use to someone if not to me.  With literal identifiers, the following would be a valid expression:
``"`Section 1-2a` + `2015 Information Table`"``
And would evaluate to the sum of whatever those two variables contain.  
I ask about the possibility, because I think it would take me a similar amount of work to add to this to expr as implementing a good workaround.

**Reply by antonmedv** [reply]
I was thinking about how to add this feature myself. But using `` ` `` is better to reserve for raw strings.

For example, in JavaScript you can use `global`:
```js
globalThis['Section 1-2a'] = 1
```

Maybe something like `env` keyword will make sense? `env['Section 1-2a']`?

Can you research how other languages are doing this?

**Reply by mdmcconnell** [reply]
Yes, I understand, and it is good to consider carefully how to use one of the last unused characters.  `env` keyword would work well, though possibly break someone's application; my only argument in favour of new syntax is slightly shorter / nicer-looking expressions.  I am not familiar with any language that allows totally arbitrary variable names.  Some [googling ](https://softwareengineering.stackexchange.com/questions/145751/has-whitespace-in-identifiers-ever-been-idiomatic)indicates Fortran ignored whitespace entirely (not really helpful here).  VB apparently allows within square brackets (very similar implementation to backtick, sort of like `env[]` with the keyword assumed; could be feasible?). [Kotlin ](https://commonsware.com/Kotlin/pages/chap-escapedmethodname-003.html)and [Scala ](https://stackoverflow.com/questions/4195848/is-there-any-language-where-names-can-include-space-characters)it seems allow using backticks, Common LISP using | and PHP may sort of using single quotes.
